Ubuntu: https://bugs.launchpad.net/ubuntu/+source/grub2/+bug/423412
Upstream: http://lists.gnu.org/archive/html/grub-devel/2009-09/msg00061.html
Description: Disable .eh_frame sections to reduce core.img size.
 Committed upstream.

diff -Nur -x '*.orig' -x '*~' grub2-1.97~beta1/configure.ac grub2-1.97~beta1.new/configure.ac
--- grub2-1.97~beta1/configure.ac	2009-08-30 17:02:11.000000000 +0100
+++ grub2-1.97~beta1.new/configure.ac	2009-09-03 16:02:49.000000000 +0100
@@ -246,6 +246,23 @@
       TARGET_CFLAGS="$TARGET_CFLAGS -malign-jumps=1 -malign-loops=1 -malign-functions=1"
     fi
   fi
+
+  # By default, GCC 4.4 generates .eh_frame sections containing unwind
+  # information in some cases where it previously did not. GRUB doesn't need
+  # these and they just use up vital space. Restore the old compiler
+  # behaviour.
+  AC_CACHE_CHECK([whether -fno-dwarf2-cfi-asm works], [grub_cv_cc_fno_dwarf2_cfi_asm], [
+    SAVE_CFLAGS="$CFLAGS"
+    CFLAGS="$CFLAGS -fno-dwarf2-cfi-asm"
+    AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[]], [[]])],
+		      [grub_cv_cc_fno_dwarf2_cfi_asm=yes],
+		      [grub_cv_cc_fno_dwarf2_cfi_asm=no])
+    CFLAGS="$SAVE_CFLAGS"
+  ])
+
+  if test "x$grub_cv_cc_fno_dwarf2_cfi_asm" = xyes; then
+    TARGET_CFLAGS="$TARGET_CFLAGS -fno-dwarf2-cfi-asm"
+  fi
 fi
 
 grub_apple_target_cc
